pkg/kademlia/pslice/pslice_test.go

```diff
@@ -0,0 +1,335 @@
+		base = test.RandomAddress()
+		a    = test.RandomAddressAt(base, 0)
+		e    = errors.New("err1")
+	)
+
+	ps.Add(a, 0)
+
+	f := func(p swarm.Address, _ uint8) (stop, jumpToNext bool, err error) {
+		return false, false, e
+	}
+
+	err := ps.EachBin(f)
+	if !errors.Is(err, e) {
+		t.Fatal("didnt get expected error")
+	}
+}
+
+// TestIterators tests that the EachBin and EachBinRev iterators work as expected.
+func TestIterators(t *testing.T) {
+	ps := pslice.New(4)
+
+	base := test.RandomAddress()
+	peers := make([]swarm.Address, 4)
+	for i := 0; i < 4; i++ {
+		a := test.RandomAddressAt(base, i)
+		peers[i] = a
+	}
+
+	testIterator(t, ps, false, false, 0, []swarm.Address{})
+	testIteratorRev(t, ps, false, false, 0, []swarm.Address{})
+
+	for i, v := range peers {
+		ps.Add(v, uint8(i))
+	}
+
+	testIterator(t, ps, false, false, 4, []swarm.Address{peers[3], peers[2], peers[1], peers[0]})
+	testIteratorRev(t, ps, false, false, 4, peers)
+
+	ps.Remove(peers[2], 2)
+	testIterator(t, ps, false, false, 3, []swarm.Address{peers[3], peers[1], peers[0]})
+	testIteratorRev(t, ps, false, false, 3, []swarm.Address{peers[0], peers[1], peers[3]})
+
+	ps.Remove(peers[0], 0)
+	testIterator(t, ps, false, false, 2, []swarm.Address{peers[3], peers[1]})
+	testIteratorRev(t, ps, false, false, 2, []swarm.Address{peers[1], peers[3]})
+
+	ps.Remove(peers[3], 3)
+	testIterator(t, ps, false, false, 1, []swarm.Address{peers[1]})
+	testIteratorRev(t, ps, false, false, 1, []swarm.Address{peers[1]})
+
+	ps.Remove(peers[1], 1)
+	testIterator(t, ps, false, false, 0, []swarm.Address{})
+	testIteratorRev(t, ps, false, false, 0, []swarm.Address{})
+}
+
+// TestIteratorsJumpStop tests that the EachBin and EachBinRev iterators jump to next bin and stop as expected.
+func TestIteratorsJumpStop(t *testing.T) {
+	ps := pslice.New(4)
+
+	base := test.RandomAddress()
+	peers := make([]swarm.Address, 12)
+	j := 0
+	for i := 0; i < 4; i++ {
+		for ii := 0; ii < 3; ii++ {
+			a := test.RandomAddressAt(base, i)
+			peers[j] = a
+			ps.Add(a, uint8(i))
+			j++
+		}
+	}
+
+	// check that jump to next bin works as expected
+	testIterator(t, ps, true, false, 4, []swarm.Address{peers[11], peers[8], peers[5], peers[2]})
+	testIteratorRev(t, ps, true, false, 4, []swarm.Address{peers[2], peers[5], peers[8], peers[11]})
+
+	// check that the stop functionality works correctly
+	testIterator(t, ps, true, true, 1, []swarm.Address{peers[11]})
+	testIteratorRev(t, ps, true, true, 1, []swarm.Address{peers[2]})
+
+}
+
+func testIteratorRev(t *testing.T, ps *pslice.PSlice, skipNext, stop bool, iterations int, peerseq []swarm.Address) {
+	t.Helper()
+	i := 0
+	f := func(p swarm.Address, po uint8) (bool, bool, error) {
+		if i == iterations {
+			t.Fatal("too many iterations!")
+		}
+		if !p.Equal(peerseq[i]) {
+			t.Errorf("got wrong peer seq from iterator")
+		}
+		i++
+		return stop, skipNext, nil
+	}
+
+	err := ps.EachBinRev(f)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if i != iterations {
```

Comment:
You don't have a matching check in `testIterator`. Is that deliberate?

Reply:
this is actually not needed since the check is already done in the callback immediately when there are too many iterations